save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/ContestEnroller.kt

```diff
@@ -0,0 +1,197 @@
+    isModalOpen: Boolean,
+    selectedContestName: String?,
+    selectedOrganizationName: String?,
+    selectedProjectName: String?,
+    onResponse: (String) -> Unit,
+    onCloseButtonPressed: () -> Unit,
+) {
+    modal { props ->
+        props.isOpen = isModalOpen
+        contestEnrollerComponent {
+            contestName = selectedContestName
+            projectName = selectedProjectName
+            organizationName = selectedOrganizationName
+            onResponseChanged = onResponse
+        }
+        div {
+            className = ClassName("d-sm-flex align-items-center justify-content-center mt-4")
+            button {
+                className = ClassName("btn btn-primary")
+                type = ButtonType.button
+                onClick = {
+                    onCloseButtonPressed()
+                }
+                +"Cancel"
+            }
+        }
+    }
+}
+
+/**
+ * A [FC] for a card that allows user to:
+ *  1. Choose a projects that should be enrolled for a contest
+ *  2. Choose a contest that the project should be enrolled for
+ *
+ * @return a functional component representing a contest enroller
+ */
+@Suppress(
+    "LongMethod",
+    "EMPTY_BLOCK_STRUCTURE_ERROR",
+    "TOO_LONG_FUNCTION",
+    "MAGIC_NUMBER",
+    "ComplexMethod",
+)
+private fun contestEnrollerComponent() = FC<ContestEnrollerProps> { props ->
+    val (organizationName, setOrganizationName) = useState(props.organizationName)
+    val (projectName, setProjectName) = useState(props.projectName)
+    val (contestName, setContestName) = useState(props.contestName)
+
+    val (availabilityUrl, _) = useState(contestName?.let {
+        "$apiUrl/contests/$it/avaliable"
+    } ?: "$apiUrl/contests/$organizationName/$projectName/avaliable")
+    val (availableOptions, setAvailableOptions) = useState(emptyList<String>())
+    useRequest(isDeferred = false) {
+        val availableVariants = get(
+            availabilityUrl,
+            headers = Headers().also {
+                it.set("Accept", "application/json")
+                it.set("Content-Type", "application/json")
+            },
+            loadingHandler = ::noopLoadingHandler,
+        )
+            .unsafeMap {
+                it.decodeFromJsonString<List<String>>()
+            }
+        setAvailableOptions(availableVariants)
+    }()
+
+    val (selectedOption, setSelectedOption) = useState("")
+    val enrollRequest = useRequest {
+        val responseFromBackend = get(
+            "$apiUrl/contests/$contestName/enroll?organizationName=$organizationName&projectName=$projectName",
+            headers = Headers(),
+            loadingHandler = ::noopLoadingHandler,
+        )
+            .text()
+            .await()
+        props.onResponseChanged(responseFromBackend)
+    }
+
+    useEffect(selectedOption) {
+        if (selectedOption.isNotBlank()) {
+            val splitResult = selectedOption.split("/")
+            if (splitResult.size == 1) {
```

Comment:
Too unclear and error-prone. I'd rather mark this more explicitly, like flag that is set at the start of render method, `isContest` or something like this. Or even make it a prop and set when creating the component

Reply:
Well it is a bit unclear, but to my mind another flag will make it even more complicated because we will need to check if all the data initialized in a correct way

Reply:
Ok, have an idea, will fix it

Reply:
Done

Reply:
Have you considered `sealed class` as a props field, that will be set to one of two variants depending on what we want to render?

Reply:
Well, it seems to be a nice idea from a first glance but you see raise of complexity while implementing

Reply:
Fixed )